src/Compilers/CSharp/Portable/Symbols/UpdatedContainingSymbolLocal.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+#nullable enable
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal sealed class UpdatedContainingSymbolAndNullableAnnotationLocal : LocalSymbol
+    {
+        internal UpdatedContainingSymbolAndNullableAnnotationLocal(SourceLocalSymbol underlyingLocal, Symbol updatedContainingSymbol, TypeWithAnnotations updatedType)
+        {
+            Debug.Assert(underlyingLocal is object);
+            Debug.Assert(updatedContainingSymbol is object);
+            UnderlyingLocal = underlyingLocal;
+            ContainingSymbol = updatedContainingSymbol;
+            TypeWithAnnotations = updatedType;
+        }
+
+        private SourceLocalSymbol UnderlyingLocal { get; }
+        public override Symbol ContainingSymbol { get; }
+        public override TypeWithAnnotations TypeWithAnnotations { get; }
+
+        public override bool Equals(Symbol other, TypeCompareKind compareKind)
+        {
+            if (other == (object)this)
+            {
+                return true;
+            }
+
+            if (!(other is LocalSymbol otherLocal))
+            {
+                return false;
+            }
+
+            SourceLocalSymbol? otherSource = otherLocal switch
+            {
+                UpdatedContainingSymbolAndNullableAnnotationLocal updated => updated.UnderlyingLocal,
+                SourceLocalSymbol source => source,
```

Comment:
>SourceLocalSymbol [](start = 16, length = 17)

Treating `UpdatedContainingSymbolLocal` and `SourceLocalSymbol` equal seems fragile.

What was the reason we chose not to clone `SourceLocalSymbol` with updated `ContainingSymbol` and `Type`?

Reply:
Cloning it would necessitate passing a binder around and potentially causing new binding to occur, and it's easier to reason about this smaller wrapper class instead of what will happen when we create a full new symbol.
Additionally, this class will likely serve as a starting point for a wrapped ITypeSymbol later when Aleksey splits the tree.

---
In reply to: [331689353](https://github.com/dotnet/roslyn/pull/38769#discussion_r331689353) [](ancestors = 331689353)